docs/architecture/adr-017-subspace-fee-grant.md

```diff
@@ -0,0 +1,261 @@
+    rpc UserAllowances(QueryUserAllowancesRequest) returns (QueryAllowancesResponse) {
+        option (google.api.http).get = "/desmos/subspaces/v3/subspaces/{subspace_id}/granter/{granter}/users/allowances";
+    }
+    // GroupAllowances returns all the grants for groups.
+    rpc GroupAllowances(QueryGroupAllowancesRequest) returns(QueryGroupAllowancesResponse) {
+        option (google.api.http).get = "/desmos/subspaces/v3/subspaces/{subspace_id}/granter/{granter}/groups/allowances";
+    }
+}
+
+// QueryUserAllowancesRequest is the request type for the Query/UserAllowances RPC method.
+message QueryUserAllowancesRequest {
+    // the id of the subspace where the granter grants the allowance to the grantee.
+    uint64 subspace_id = 1;
+
+    // the address of the user granting an allowance of their funds.
+    string granter = 2;
+
+    // (Optional) the address of the user being granted an allowance of another user's funds.
+    string grantee = 3;
+
+    // pagination defines an pagination for the request.
+    cosmos.base.query.v1beta1.PageRequest pagination = 4;
+}
+
+// QueryUserAllowancesResponse is the response type for the Query/UserAllowances RPC method.
+message QueryUserAllowancesResponse {
+    // allowances are allowance's granted for grantee by granter.
+    repeated cosmos.feegrant.v1beta1.Grant allowances = 1;
+
+    // pagination defines an pagination for the response.
+    cosmos.base.query.v1beta1.PageResponse pagination = 2;
+}
+
+// QueryGroupAllowancesRequest is the request type for the Query/GroupAllowances RPC method.
+message QueryGroupAllowancesRequest {
+    // the id of the subspace where the granter grants the allowance to the grantee.
+    uint64 subspace_id = 1;
+
+    // the address of the user granting an allowance of their funds.
+    string granter = 2;
+
+    // (Optional) the id of the group being granted an allowance of another user's funds.
+    uint32 group_id = 3;
+
+    // pagination defines an pagination for the request.
+    cosmos.base.query.v1beta1.PageRequest pagination = 4;
+}
+
+// QueryGroupAllowancesResponse is the response type for the Query/GroupAllowances RPC method.
+message QueryGroupAllowancesResponse {
+    // allowances are allowance's granted for grantee by granter.
+    repeated cosmos.subspace.v3.GroupGrant allowances = 1;
+
+    // pagination defines an pagination for the response.
+    cosmos.base.query.v1beta1.PageResponse pagination = 2;
+}
+```
+
+## Consequences
+
+### Backwards Compatibility
+
+The proposed solution introduces a new set of store keys, it is completely backward compatible.
+
+### Positive
+
+- Allow subspace fee providers to grant fees allowances to their users.
+
+### Negative
+
+- Storing extra subspace grant info takes up more storage space;
+- Having more check in the deducting fees phase slow down the transaction operation;
+- Removing expired allowances at the block begin phase increases the block production time.  
```

Comment:
Is this necessary? Does the Cosmos SDK remove the fee grants at each block, or do they remove them when they are tried to be executed, if no longer valid?

Reply:
Cosmos SDK remove all the expired fee grants every block. We can make it removed when the user try to execute with feegrant, but it will leave invalid allowances in the database. I prefer to remove expired allowances every blocks.

Reply:
Ok, then I think we should add this inside the ADR since there's not reference to a `BeginBlock` implementation